bigquery_etl/view/validate_views.py

```diff
@@ -0,0 +1,134 @@
+    "sql/moz-fx-data-shared-prod/search/search_aggregates/view.sql",
+    "sql/moz-fx-data-shared-prod/telemetry/clients_probe_processes/view.sql",
+}
+
+parser = ArgumentParser(description=__doc__)
+parser.add_argument(
+    "--sql_dir",
+    "--sql-dir",
+    help="Directories with view definitions to be validated.",
+    default="sql/",
+)
+standard_args.add_log_level(parser)
+
+
+def validate_view_naming(view_file):
+    """Validate that the created view naming matches the directory structure."""
+    with open(view_file) as f:
+        sql = f.read()
+    parsed = sqlparse.parse(sql)[0]
+    tokens = [
+        t
+        for t in parsed.tokens
+        if not (t.is_whitespace or isinstance(t, sqlparse.sql.Comment))
+    ]
+    is_view_statement = (
+        " ".join(tokens[0].normalized.split()) == "CREATE OR REPLACE"
+        and tokens[1].normalized == "VIEW"
+    )
+    if is_view_statement:
+        target_view = str(tokens[2]).strip().split()[0]
+        try:
+            [project_id, dataset_id, view_id] = target_view.replace("`", "").split(".")
+            if not (
+                view_file.parent.name == view_id
+                and view_file.parent.parent.name == dataset_id
+                and view_file.parent.parent.parent.name == project_id
+            ):
+                print(
+                    f"{view_file} ERROR\n"
+                    f"View name {target_view} not matching directory structure."
+                )
+                return False
+        except Exception:
+            print(f"{view_file} ERROR\n{target_view} missing project ID qualifier.")
+            return False
+    else:
+        print(
+            f"ERROR: {view_file} does not appear to be "
+            "a CREATE OR REPLACE VIEW statement! Quitting..."
+        )
+        return False
+    return True
+
+
+def validate_fully_qualified_references(view_file):
+    """Check that referenced tables and views are fully qualified."""
+    with open(view_file) as f:
+        sql = f.read()
+
+    # dry run only returns referenced tables, not views
+    referenced_tables = DryRun(str(view_file)).get_referenced_tables()
+    for line in sql.split("\n"):
+        for referenced_table in referenced_tables:
+            # If a view is referenced in the query then instead of the view name,
+            # the tables that are referenced it the view definition are returned.
+            # Some of these tables are not part of the SQL of the view that is
+            # validated.
+            ref_project = referenced_table[0]
+            ref_dataset = referenced_table[1]
+            ref_table = referenced_table[2]
+            if f"{ref_dataset}.{ref_table}" in line:
+                if f"{ref_project}.{ref_dataset}.{ref_table}" not in line:
+                    print(
+                        f"{view_file} ERROR\n"
+                        f"{ref_dataset}.{ref_table} missing project ID qualifier."
+                    )
+                    return False
```

Comment:
One issue with this is that this won't detect cases where a view definition references another view that is not fully qualified. Dry run only seems to return tables, not views :/

Reply:
We could avoid checking these references ourselves if we dry run views in a different project that doesn't contain any BigQuery datasets.